src/api/app/views/webui/projects/maintained_projects/index.html.haml

```diff
@@ -0,0 +1,39 @@
+:ruby
+  @pagetitle = "Projects Maintained by #{@project}"
+
+.card
+  .card-body
+    - if policy(@project).update?
+      %ul.list-inline
+        %li.list-inline-item.mr-3
+          = link_to('#', data: { toggle: 'modal', target: '#new-maintenance-project-modal' }) do
+            %i.fas.fa-plus-circle.text-primary
+            Add New Project
+    %h3 Maintained Projects
+    .table-responsive
+      %table.table.table-sm.table-striped.table-bordered
+        %thead
+          %tr
+            %th Name
+            %th Actions
+        %tbody
+          - @maintained_projects.each do |project_name|
+            %tr
+              %td
+                = link_to(project_name, project_show_path(project_name))
+              %td
+                - if policy(@project).update?
+                  = link_to('#', title: 'Delete Project',
+                                 data: { toggle: 'modal', target: '#delete-maintained-project-modal',
+                                         action: project_maintained_project_path(project_name: @project, maintained_project: project_name) }) do
+                    %i.fas.fa-times-circle.text-danger
+
+- if policy(@project).update?
```

Comment:
@coolo I don't get it why the "number of lines" are important here. So if we don't use partials have 300 lines, would be ok? BTW In the original it is being done via ```User.current.can_modify?(@project)``` which I don't think it's a problem either (since we have cache query). Enabling caching on development with ```config.action_controller.perform_caching = true``` makes possible to see that the DB just get hits once. I will try it with a higher number of Projects and I will let you know.

Reply:
the number of lines in between the calls matters - because it's making it so damn obvious just how wasteful this type of programming is. Checking the DB query cache is effort - and this effort can be so easily avoided by putting function call results into a variable. This is not ruby specific

Reply:
Make it work, make it right, make it fast. Can you show me some numbers in the production environment setup that justify make early optimizations?

Reply:
Dude, this is not early optimization, this is responsibe programming following basic rules.

Reply:
@coolo, the usage of the instance variable, didn't save me a second, while running it, in production, having the following setup:

```
projects = 0.upto(100000).collect do |p|
	pname = "openSUSE:Leap:Test:#{p}"
	Project.create(name: pname)
end

mname = "openSUSE:Maintenance"
maintenance_project = Project.create!(name: mname, kind: "maintenance")

projects.each do |prj|
		MaintainedProject.create(project: prj, maintenance_project: maintenance_project)
end

```

then try to access the 

http://localhost:3000/projects/openSUSE:Maintenance/maintained_projects

and it hangs on the policy, because of the flags lookup on ```Policy#is_locked?``` method. If we bring me to the question, wouldn't be better to have a lighter policy to the views where we are just trying to hide the icon from the user? @hennevogel

Reply:
lighter policy?

Reply:
Aka the 'what can go wrong if we didn't read the AGB beyond the title?' policy :)

Reply:
> lighter policy?

i.e in the webui we are just concerned in hide/show the icons, so for example, it would be much faster, and "almost always right :)" if we had just a ```User.current.can_modify_project?(@project, false)``` in a policy, just used by the views. In the action we still have a full policy in place. 

- Full policy for controllers, used with ```authorize```
- Light policy for the view used with ```policy``` (to control displayed UI elements)

Reply:
So by "light" you mean "not as accurate". How can we achieve light as in fast **and** accurate? :)

Reply:
> So by "light" you mean "not as accurate". How can we achieve light as in fast **and** accurate? :)

It is accurate, not 100% of the time :)

Reply:
the only scenario that we miss is if the project is "locked"

Reply:
Just don't repeat the check for @project and you're good

Reply:
@coolo please read it https://github.com/openSUSE/open-build-service/pull/7379#discussion_r273940768

Reply:
OK, I tried and I'm right :)

```
--- a/src/api/app/views/webui/projects/maintained_projects/index.html.haml
+++ b/src/api/app/views/webui/projects/maintained_projects/index.html.haml
@@ -1,9 +1,10 @@
 :ruby
   @pagetitle = "Projects Maintained by #{@project}"
+  can_update_project = policy(@project).update?  
 
 .card
   .card-body
-    - if policy(@project).update?
+    - if can_update_project
       %ul.list-inline
         %li.list-inline-item.mr-3
           = link_to('#', data: { toggle: 'modal', target: '#new-maintenance-project-modal' }) do
@@ -22,13 +23,13 @@
               %td
                 = link_to(project_name, project_show_path(project_name))
               %td
-                - if policy(@project).update?
+                - if can_update_project
                   = link_to('#', title: 'Delete Project',
                                  data: { toggle: 'modal', target: '#delete-maintained-project-modal',
                                          action: project_maintained_project_path(project_name: @project, maintained_project: project_name) }) do
                     %i.fas.fa-times-circle.text-danger
 
-- if policy(@project).update?
+- if can_update_project
   = render partial: 'delete_dialog'
   = render partial: 'create_dialog'
```

results in only 1 query for 12 maintained projects as shown here:

```
frontend_1  | 14:22:52 clock.1       | I, [2019-04-10T14:22:52.001185 #9]  INFO -- : Triggering 'fetch notifications'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.93] Started GET "/projects/openSUSE:Maintenance/maintained_projects" for 10.160.66.102 at 2019-04-10 14:22:53 +0000
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.94] Processing by Webui::Projects::MaintainedProjectsController#index as HTML
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.94]   Parameters: {"project_name"=>"openSUSE:Maintenance"}
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   User Load (0.3ms)  SELECT  `users`.* FROM `users` WHERE `users`.`login` = 'coolo' LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   Configuration Load (0.4ms)  SELECT  `configurations`.* FROM `configurations` ORDER BY `configurations`.`id` ASC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   CACHE Configuration Load (0.0ms)  SELECT  `configurations`.* FROM `configurations` ORDER BY `configurations`.`id` ASC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   CACHE Configuration Load (0.0ms)  SELECT  `configurations`.* FROM `configurations` ORDER BY `configurations`.`id` ASC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   CACHE Configuration Load (0.0ms)  SELECT  `configurations`.* FROM `configurations` ORDER BY `configurations`.`id` ASC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   Announcement Load (0.2ms)  SELECT  `announcements`.* FROM `announcements` ORDER BY `announcements`.`created_at` DESC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   CACHE Announcement Load (0.0ms)  SELECT  `announcements`.* FROM `announcements` ORDER BY `announcements`.`created_at` DESC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.95]   Role Exists (0.4ms)  SELECT  1 AS one FROM `roles` INNER JOIN `roles_users` ON `roles`.`id` = `roles_users`.`role_id` WHERE `roles_users`.`user_id` = 138 AND `roles`.`title` = 'Admin' LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.96]   Project Load (0.4ms)  SELECT  `projects`.* FROM `projects` WHERE (projects.id not in (50315,60546,69200,122039,122041,122042,122043,189310,248287,254068,97989,277778)) AND `projects`.`name` = 'openSUSE:Maintenance' LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.96]    (0.5ms)  SELECT projects.name FROM `maintained_projects` LEFT OUTER JOIN `projects` ON `projects`.`id` = `maintained_projects`.`project_id` AND (projects.id not in (50315,60546,69200,122039,122041,122042,122043,189310,248287,254068,97989,277778)) WHERE `maintained_projects`.`maintenance_project_id` = 43794
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.96]   Rendering webui/projects/maintained_projects/index.html.haml within layouts/webui2/webui
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]   Flag Exists (0.4ms)  SELECT  1 AS one FROM `flags` WHERE `flags`.`project_id` = 43794 AND `flags`.`flag` = 'lock' AND `flags`.`status` = 'enable' LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97] has_global_permission? create_package
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]   Role Load (0.5ms)  SELECT DISTINCT `roles`.* FROM `roles` INNER JOIN `roles_users` ON `roles`.`id` = `roles_users`.`role_id` WHERE `roles_users`.`user_id` = 138
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]   StaticPermission Load (0.4ms)  SELECT  DISTINCT `static_permissions`.* FROM `static_permissions` INNER JOIN `roles_static_permissions` ON `static_permissions`.`id` = `roles_static_permissions`.`static_permission_id` WHERE `roles_static_permissions`.`role_id` = 2 AND `static_permissions`.`title` = 'create_package' LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]    (0.4ms)  SELECT `roles_static_permissions`.`role_id` FROM `roles_static_permissions` INNER JOIN `static_permissions` ON `static_permissions`.`id` = `roles_static_permissions`.`static_permission_id` WHERE `static_permissions`.`title` = 'create_package'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97] running local permission check: user coolo, project openSUSE:Maintenance, permission 'create_package'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]   Project Load (0.3ms)  SELECT  `projects`.* FROM `projects` WHERE (projects.id not in (50315,60546,69200,122039,122041,122042,122043,189310,248287,254068,97989,277778)) AND `projects`.`name` = 'openSUSE' ORDER BY `projects`.`name` DESC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.97]   Relationship Exists (0.3ms)  SELECT  1 AS one FROM `relationships` WHERE `relationships`.`project_id` = 43794 AND `relationships`.`user_id` = 138 AND (role_id in (1,3)) LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   Relationship Exists (0.5ms)  SELECT  1 AS one FROM `relationships` INNER JOIN `groups` ON `groups`.`id` = `relationships`.`group_id` INNER JOIN `groups_users` ON `groups_users`.`group_id` = `groups`.`id` WHERE `relationships`.`project_id` = 43794 AND `groups_users`.`user_id` = 138 AND (role_id in (1,3)) LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98] permission not found, trying parent project 'openSUSE'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   CACHE  (0.0ms)  SELECT `roles_static_permissions`.`role_id` FROM `roles_static_permissions` INNER JOIN `static_permissions` ON `static_permissions`.`id` = `roles_static_permissions`.`static_permission_id` WHERE `static_permissions`.`title` = 'create_package'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98] running local permission check: user coolo, project openSUSE, permission 'create_package'
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   Project Load (0.3ms)  SELECT  `projects`.* FROM `projects` WHERE (projects.id not in (50315,60546,69200,122039,122041,122042,122043,189310,248287,254068,97989,277778)) AND 1=0 ORDER BY `projects`.`name` DESC LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   Relationship Exists (0.4ms)  SELECT  1 AS one FROM `relationships` WHERE `relationships`.`project_id` = 23185 AND `relationships`.`user_id` = 138 AND (role_id in (1,3)) LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   Relationship Exists (0.3ms)  SELECT  1 AS one FROM `relationships` INNER JOIN `groups` ON `groups`.`id` = `relationships`.`group_id` INNER JOIN `groups_users` ON `groups_users`.`group_id` = `groups`.`id` WHERE `relationships`.`project_id` = 23185 AND `groups_users`.`user_id` = 138 AND (role_id in (1,3)) LIMIT 1
frontend_1  | 14:22:53 web.1         | [cd0ca564-0fb3-4092-933d-2ae0817249da] [7:415.98]   Rendered webui/projects/maintained_projects/index.html.haml within layouts/webui2/webui (22.2ms)
```

Reply:
And no, it doesn't save a second. But your code: 339ms, my code 229ms.

From that I conclude one policy() call is 8.4ms - which means you need 119 maintained projects to save a second. You can do it! :)

Reply:
wow 8.4 ms? for 50.000 projects it doesn't work either. For 10 projects it doesn't make difference..

Reply:
You don't call 100ms a difference? If the policy takes 8.4ms or 4ms or 15ms doesn't make a difference - unless you call it 13 times too often (for 12 projects). That's what I'm trying to say. But if it doesn't help with 50.000 projects, we shall have a look why.

Reply:
@coolo the first try will be to use the server side pagination and hope that we can cache there the policy call